Adding an abstraction for log segments

vgkholla opened the pull request:
This commit adds the abstraction for representing log segments. Log segments will
form the components of a larger log abstraction. For design and discussion please
refer to https://github.com/linkedin/ambry/issues/469.

Please use the current implementation of [`Log`](https://github.com/linkedin/ambry/blob/e85e3a2fbf190ceca91a8f808663f7bf0b195ed7/ambry-store/src/main/java/com.github.ambry.store/Log.java) as a reference.

Built (`./gradlew build && ./gradlew test`) and formatted.

**Primary reviewers: @pnarayanan, @nsivabalan**
**Expected time to review: 60-90 mins**

**Unit testing**

Class | Class, % | Method, % | Line, %
------- | ------------ | -------------- | ----------
`LogSegment` | 100% (2/ 2) | 100% (17/ 17) | 100% (94/ 94)

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * A LogSegment once instantiated can only go from FREE -> ACTIVE -> SEALED in the natural course of operations, never
+   * backward. However it can be externally manipulated to move in any direction.
+   */
+  State state;
```
why state and nextSegment not private?

vgkholla replied:
Because they can be set/reset in recovery code - just like `endOffset`

nsivabalan replied:
ok fine. Can we have methods to set and get instead of accessing it directly

vgkholla replied:
I don't see why - we would need both a setter and getter that simply set/get the value without any extra processing.

nsivabalan commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -0,0 +1,702 @@
+      currSegment.writeFrom(Channels.newChannel(new ByteBufferInputStream(firstBuf)), currSegment.getEndOffset(),
+          firstWriteSize);
+      assertEquals("End offset in current segment is not equal to the cumulative bytes written", firstWriteSize,
+          currSegment.getEndOffset());
```
can we also check the state of the log segment as well

vgkholla replied:
That happens as a part of other tests.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * Used to describe the state of the LogSegment.
+   */
+  enum State {
+    FREE,
```
Sure. Will add.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+  /**
+   * The state of the segment.
+   * <p/>
+   * A LogSegment once instantiated can only go from FREE -> ACTIVE -> SEALED in the natural course of operations, never
```
The file is recycled - the `LogSegment` instance is not. Even if it will be, it is not the "normal" course of operations. Let me know if you think this can be made clearer.

nsivabalan replied:
got you

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   */
+  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics)
+      throws IOException {
+    if (!file.exists() || !file.isFile()) {
```
No - the file will have been preallocated before creating a `LogSegment` instance to represent it.

nsivabalan replied:
Why is it so? If we consider the current Log, it pre-allocates if not available. Was there any issue with that ?

vgkholla replied:
No issue - a higher level entity (the `Log` that encompasses these `LogSegment` instances) will now do the preallocation. This is to allow pooling of segments - might become clearer once the `Log` and `DiskManager` are ready.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+    } else if (endOffset.get() + buffer.remaining() > capacityInBytes) {
+      // maintain the invariant that there are no "partial" writes. A log segment should write the data that it receives
+      // completely and the data that was provided in a single call cannot be written across multiple segments.
+      logger.info("{} : Writing data of size {} from a buffer. Current end offset is {} and capacity is {}",
```
will add a line talking about rollover.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * If {@code offset} + {@code size} exceeds the capacity of the current segment, bytes are written starting from
```
Not required - looked like an easy (and right) thing to have.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * <p/>
+   * If {@code position} + {@code buffer.remaining()} exceeds the capacity of the current segment, bytes are read
+   * starting from {@code position} until capacity is reached. The rest of the bytes upto {@code buffer.remaining()} are
+   * read from the next segment (if one exists). A read is not started if it cannot be completed.
```
Same reply.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * Gets the {@link File} and {@link FileChannel} backing this log segment. Also increments a ref count.
+   * <p/>
+   * The expectation is that a matching {@link #closeView()} will be called eventually to decrement the ref count.
+   * @param offset the offset that will used to start the read operation from.
```
Good points. This function is built to provide the same guarantees as the `getView()` function in `Log` (which takes in an offset). The offset is rechecked in `StoreMessageReadSet`. 

This function is a package private function. The assumption is that only components that know the guarantees of this function use it. The purpose of the offset is to do some initial checking (similar to the same function in the `Log`).

nsivabalan replied:
If we are trying to mimic getView() in current Log, then isn't the api should look like getView(List<BlobReadOptions>). But then looks like you have designed the api with the intention to create the StoreMessageReadSet in Log and not within the logSegment per se. So, I don't see a need for an offset here. Anyways, lets try to sync up offline today regarding this.

pnarayanan replied:
@vgkholla Could you note here as to why you think the `offset` is needed here? Would also be worthwhile updating the javadoc about this - "offset that will be used to start the read from" - what read?

vgkholla replied:
Added a doc. Let me know if you think it makes sense + makes it cleaer

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   */
+  Pair<File, FileChannel> getView(long offset) {
+    if (offset < 0 || offset > getEndOffset()) {
+      throw new IllegalArgumentException("Provided offset [" + offset + "] is out of segment range");
```
Sure will do.

vgkholla replied:
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+  private void ensureNext(Counter errorCounter) {
+    if (next == null) {
+      errorCounter.inc();
+      throw new IllegalStateException(file.getAbsolutePath() + ": Cannot perform operation beyond total capacity");
```
Then as far as the segment (and the higher log abstraction) is concerned, the total capacity has been reached.

nsivabalan replied:
ok got you

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -0,0 +1,702 @@
+          segment.getEndOffset(), writeSize);
+      assertEquals("End offset is not equal to the cumulative bytes written", 3 * writeSize, segment.getEndOffset());
+
+      readAndEnsureMatch(segment, 0, buf);
```
Sure.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -0,0 +1,702 @@
+      segment = new LogSegment(segmentName, new File(tempDirName, segmentName), STANDARD_SEGMENT_SIZE, metrics);
+      segment.state = LogSegment.State.ACTIVE;
+      segment.setEndOffset(buf.length);
+      readAndEnsureMatch(segment, 0, buf);
```
This one does not require multiple calls. It is only checking if the data is still around.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+ * number of the segment. Each segment knows the segment that "follows" it logically (if such a segment exists) and can
+ * transparently redirect operations if required. A segment can be in one of the states represented by {@link State}.
+ */
+class LogSegment implements Read, Write {
```
Why is an interface required? No external components (outside of the store) use it and there is no need to have different implementations at the same time that can be switched. This is similar to the `Log` which is also not an interface.

vgkholla replied:
The `Log`, `Index`, `Journal` and `BlobStore` implementations are tightly coupled with each other. Having an interface does not mean that the `Log` implementation can be changed without the implementations of the others i.e. its not a component you can just use transparently without knowing the implementation. 

If someone needs to change the `LogSegment` implementation, it is easy to do so - I see no need for multiple implementations of `LogSegment`. The `Store` is the level of abstraction that actually changes i.e. you can have a different kind of store.

vgkholla replied:
Yup. The `Store` is the interface.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -0,0 +1,702 @@
+      assertEquals("End offset is not equal to the cumulative bytes written", 3 * writeSize, segment.getEndOffset());
+
+      readAndEnsureMatch(segment, 0, buf);
+      // test that file size and end offset
```
will fix

nsivabalan replied:
yep, I knew. that was my intention too

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * <p/>
+   * If {@code offset} + {@code size} exceeds the capacity of the current segment, bytes are written starting from
+   * {@code offset} until capacity is reached. The rest of the bytes upto {@code size} are forwarded to the next segment
+   * (if one exists). A write is not started if it cannot be completed.
```
Removed.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+   *
+   */
+  @Override
+  public void writeFrom(ReadableByteChannel channel, long offset, long size)
```
whatever is the purpose, again I think these semantics will be cleaner if the caller manages the write (so it can ensure that "offset + size" is not greater than capacity and so on).

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,363 @@
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
```
Since there could be many threads accessing the log segment like actual store for regular puts and gets, compaction, may be scrubber and so on, I guess to think about thread safety. We didn't face any issue with our existing Log was that, we had locks in BlobStore during such access to Log.

vgkholla replied:
Yes and those locks will continue. The atomicity required is both the log and index together for most operations (like store put and delete) - it will have to done under a lock anyway. 

Gets still don't have to take a lock and compaction is designed to avoid locking. Trying to make this thread safe only complicates the code without any extra benefits is what I think.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+
+  /**
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
```
The filename can contain extra things not relevant to actual name of the log segment. Some examples

1) log_current -> empty name. This is required to backwards compatible with existing segments.
2) pos_gen_log_temp -> during compaction this is the filename, but the name of the segment doesn't need the "_temp".

vgkholla replied:
Not sure what you want the doc to say. The `LogSegment` doesn't care whether the name and filename are different or not - it provides flexibility. I don't want to give examples of filenames and names here because they may evolve and change and the docs will have to be updated.

pnarayanan replied:
The name by itself is not a very clear concept for the `LogSegment`. If it is not tightly coupled with a backing file, what does it really signify? Is it used to distinguish a `LogSegment` from another - as in is it an "id"? Is it the name of the file that this `LogSegment` will be eventually persisted as? Regardless of how things evolve, I think you need to clarify the significance of a `name`.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
+   * @param file the backing file for this segment.
+   * @param capacityInBytes the intended capacity of the segment
```
This is how the `Log` also does it. Detecting the size of a preallocated file is unfortunately not supported in java. It needs to be written into the segment itself or in some other descriptor. The caller has to make sure that the file and its capacity match. I see no other reliable way of doing this.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+      endOffset.addAndGet(bytesWritten);
+    }
+    return bytesWritten;
+  }
```
The `LogSegment` `next` determines ordering. Nobody else (needs to) stores the ordering.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Guarantees that the {@code buffer} is either written in its entirety in this segment or is transparently forwarded
```
Will do.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if there is not enough space for {@code buffer} and forwarding is not possible
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
```
Removed

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
```
As discussed, this is due to the fact that this is passed as a `Write` interface to the `MessageFormartWriteSet` and has to transparently be able to accommodate all the writes even though the current segment's capacity is full.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+              + "Current end offset is {} and capacity is {}", file.getAbsolutePath(), next.getName(),
+          buffer.remaining(), getEndOffset(), capacityInBytes);
+      state = State.SEALED;
+      bytesWritten = next.appendFrom(buffer);
```
Done.

vgkholla commented:
@nsivabalan @pnarayanan Addressed comments.

vgkholla commented:
@nsivabalan @pnarayanan Removed state change logic out `LogSegment` as suggested. Please take a look.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+    ensureActive();
+    if (endOffset.get() + buffer.remaining() > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
```
why IllegalArgException? I feel that IllegalStateException might be a good choice for this case.

vgkholla replied:
I don't think this is easy to call - my thinking is that this is an illegal argument because the segment exposes how much capacity it has left and if someone sent in a buffer of a higher size, it is probably an illegal argument? Anyway, not inclined either way - if you feel strongly about it, can change.

nsivabalan replied:
@pnarayanan : do you have any comments on this?

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * {@inheritDoc}
+   * <p/>
+   * The read is not started if it cannot be completed.
+   * @param buffer The buffer into which the read needs to write to
```
minor: "needs to be written to"

vgkholla replied:
this is derived from the `Read` interface where it makes sense. I have changed this to "The buffer into which the data needs to be written"

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * @throws IOException if data could not be written to the file because of I/O errors
+   *
+   */
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
```
Can this be allowed in an ACTIVE segment. I thought its allowed only on a SEALED one. Are these to accommodate zeroing out the bytes when the segment moves from SEALED to FREE state ?

nsivabalan replied:
got you

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param channel The channel from which data needs to be written from.
+   * @param offset The offset in the segment at which tho start writing.
```
typo "which to start"

vgkholla replied:
Fixed.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+    }
+    if (offset + size > endOffset.get()) {
+      endOffset.set(offset + size);
```
depending on your answer to my previous question, might have a question wrt updating end offset

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   */
+  private void ensureActive() {
+    if (!state.equals(State.ACTIVE)) {
+      throw new IllegalStateException("Segment is not in ACTIVE state");
```
minor grammatical error. "not in an ACTIVE"

vgkholla replied:
Changed.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+      throw new IllegalArgumentException(file.getAbsolutePath() + " does not exist or is not a file");
+    }
+    this.file = file;
+    this.name = name;
```
minor: Should we have a validator for the name that it follows certain format ?

vgkholla replied:
The log segment doesn't care about the name - the log will pick up the right files so there is no validation required here.

nsivabalan replied:
got you

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+
+  /**
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
```
The log segments themselves don't care about the name - that is only important to the log and so will come with the log.

nsivabalan replied:
Fine

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+    this.name = name;
+    this.capacityInBytes = capacityInBytes;
+    this.metrics = metrics;
+    fileChannel = Utils.openChannel(file, true);
```
Compaction might require more arguments (like direct I/O or not). I don't want to add those arguments until there is complete clarity - that will happen once the log segmentation impl is done.

nsivabalan commented:
Looks good except for some minor comments.

vgkholla commented:
@nsivabalan Addressed those comments.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+    while (bytesWritten < size) {
+      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+    }
+    if (offset + size > endOffset.get()) {
```
couldn't this potentially create holes in the file, if the offset provided is greater than the current `endOffset` to start with? Are you banking on the caller doing the right thing? Does the existing `Log` do this the same way?

vgkholla replied:
Yeah - banking on the caller to do the right things here since this is a specialized package private function. I do not want to limit options and tailor to one use case (for e.g. hard delete).

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * @throws IllegalArgumentException if {@code offset} < 0 or is > {@link #getEndOffset()}.
+   */
+  Pair<File, FileChannel> getView(long offset) {
+    if (offset < 0 || offset > getEndOffset()) {
```
I don't understand what `offset` will be passed to this method, but I think you want to do `offset >= getEndOffset()` here (for the reasons I mentioned earlier).

vgkholla replied:
The offset which will be eventually read. Made the change.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+
+  /**
+   * Sets the end offset of this segment. This can be lesser than the actual size of the file and represents the offset
+   * until which data that is readable is stored.
```
I like to look at this offset as the start offset for an append, as I mentioned earlier - just due to the off-by-1 aspect. You might want to clarify it here.

vgkholla replied:
Done.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+  void setEndOffset(long endOffset)
+      throws IOException {
+    long fileSize = sizeInBytes();
+    if (endOffset < 0 || endOffset > fileSize) {
```
I am unclear as to why `writeFrom()` is allowed to indirectly set the `endOffset` at arbitrary positions.

vgkholla replied:
It may create holes and it might push the end offset farther but as mentioned earlier, I expect the caller to know what is happening - I do not want to limit the flexibility of the function and tailor it to a single use case.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -119,8 +119,7 @@ public void appendFrom(ReadableByteChannel channel, long size)
 
-  @Override
-  public void writeFrom(ReadableByteChannel channel, long offset, long size)
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
```
Done. This is going to be removed in the next PR so that documentation is minimal.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+  /**
+   * Reference to the segment that logically "follows" this segment.
+   */
+  LogSegment next = null;
```
no, this isn't being used right now. The log will need to manipulate this and I didn't see any reason for adding getter/setter because they wouldn't do anything special. I don't see a reason to not allow components to set it directly.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+  /**
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
+   * @param file the backing file for this segment.
```
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
```
Yeah - it is only used in tests and I'm all for removing it. Created an issue (#476) for this.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public void readInto(ByteBuffer buffer, long position)
```
Added documentation in the `Read` interface.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+  @Override
+  public void readInto(ByteBuffer buffer, long position)
+      throws IOException {
+    if (position < 0 || position > getEndOffset()) {
```
Done.

vgkholla replied:
Also added a test.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+      throws IOException {
+    if (position < 0 || position > getEndOffset()) {
+      throw new IllegalArgumentException(
+          "Provided position [" + position + "] is out of bounds for the segment [" + file.getAbsolutePath()
```
Changed to `IndexOutOfBoundsException`

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   */
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
+      throws IOException {
+    if (offset < 0 || offset > capacityInBytes) {
```
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,391 @@
+    /**
+     * The LogSegment is completely free and no writes have been accommodated in the segment.
+     */
+    FREE,
```
I removed the concept of `State`. It is no longer useful since the `LogSegment` does not have forwarding responsibility anymore. The `Log` can easily maintain a reference to a segment is "active" since there only be one.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,339 @@
+   * The natural state transition of a LogSegment is FREE -> ACTIVE -> SEALED but it can be externally manipulated to
+   * move in any direction.
+   */
+  State state;
```
removed the concept of state

vgkholla commented:
@pnarayanan Addressed all comments.
@nsivabalan the concept of log segment state has been removed (because it does not make sense anymore). You might want to have quick look at the two most recent commits.

nsivabalan commented:
So, we are going to rely on the Log or the caller to ensure we don't write to a sealed segment or start writing to an active segment assuming it was a free segment is it ?

vgkholla commented:
the `Log`, yes - since it knows which segment is "active"

nsivabalan commented:
@vgkholla : I get the over all picture. But then, log segments are not used only by Log alone. We will have compaction threads dealing with log segments, hard deletes directly dealing with compactions. So, how are we going to guard against mishandling from those callers. From what I could think of is that, those log segments will be given out by Log and hence they (compactor) are not allowed to step into any other log segments. But shouldn't the LogSegment themselves have guards against such mis-handling?

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -0,0 +1,303 @@
+  /**
+   * Reference to the segment that logically "follows" this segment.
+   */
+  LogSegment next = null;
```
Again, what do you need `next` for? The order is now maintained outside of this segment, isn't it? Can't you remove this altogether? (sorry if I missed your reply - can't see it here).

vgkholla commented:
@nsivabalan Compaction will also use the `Log` abstraction (more details in the design document and the gist) and not deal with log segments. Hard Delete uses `writeFrom()` which does not do state checking anyway - so I think there is no problem.

pnarayanan commented:
LGTM
